### PR TITLE
Update emailing service to sendgrid

### PR DESCRIPTION
**Pull request by Jood80**
## In This PR:

- [x] switch from [nodemailer](https://nodemailer.com/about/) to [sendgird](https://sendgrid.com/) service.
- [x] update email templates.

* **effected endpoints**:
    1.  `POST`: `api/v1/auth/signup/` - as User
    2. `POST`: `api/v1/auth/login/` - both User & Admin 
    3. `GET`: `api/v1/user/waiting-list/` - as Admin
    4. `POST`: `api/v1/user/approve/:userId` - as Admin
    5. `POST`: `api/v1/user/reject/:userId` - as Admin
    6. `POST`: `api/v1/auth/forget-password` - as User
    7. `POST`:`api/v1/auth/reset-password` - as a User. 
---
### Pricing:
- with our free plan now, we can send up to 100 emails/day.
- exceeding that free amount, to send between [1000 - 50,000] email/month, will cost us/them ~$19.95/monthly.
for more [see pricing section](https://sendgrid.com/pricing/).
-----
### Email Templates:
- [x] the user should be expecting 4 different email types from us:
  - Verification email
  - Approval Email
  - Rejection Email
  - Reset password email
<div align="center">

</div>

----
#### Env Config
1. setup your backend environment.
   -  <details>
      <summary>create `.env.development` with these variables.</summary>
      
      ```
        PORT=5000
        DATABASE_URL = postgres://<USERNAME>:<PASSWORD>@localhost:5432/nextup_comedy_dev
        TEST_DB_URL = postgres://<USERNAME>:<PASSWORD>@localhost:5432/nextup_comedy_test
        JWT_SECRET=<IT_IS_CALLED_SECRET_FOR_A_REASON>
        NODEMAILER_APP_MAIL = <IT_IS_CALLED_SECRET_FOR_A_REASON>
        NODEMAILER_MAIL_PASSWORD = <IT_IS_CALLED_SECRET_FOR_A_REASON>
        CLIENT_BASE_URL = <IT_IS_CALLED_SECRET_FOR_A_REASON>
        SERVER_BASE_URL = <IT_IS_CALLED_SECRET_FOR_A_REASON>
        
        SENDGRID_ADMIN_EMAIL = <IT_IS_CALLED_SECRET_FOR_A_REASON>
        SENDGRID_API_KEY = <IT_IS_CALLED_SECRET_FOR_A_REASON>
        SENDGRID_VERIFICATION_TEMPLATE_ID = <IT_IS_CALLED_SECRET_FOR_A_REASON>
        SENDGRID_RESET_PASSWORD_TEMPLATE_ID = <IT_IS_CALLED_SECRET_FOR_A_REASON>
        SENDGRID_APPROVAL_TEMPLATE_ID = <IT_IS_CALLED_SECRET_FOR_A_REASON>
        SENDGRID_REJECTION_TEMPLATE_ID = <IT_IS_CALLED_SECRET_FOR_A_REASON>
        NEXTUP_COMEDY_SUPPORT_EMAIL = <IT_IS_CALLED_SECRET_FOR_A_REASON>
        
        CLIENT_ID=XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXx
        CLIENT_SECRET=XXXXXXXXXXXXXXXXx
      ```
      </details>
  - `$ npm run build:db` >`$ npm run insert:db`
  - `$ npm install `:
     -  `$ npm run dev`
     - OR : create `.env` > `$ npm run build` > `$ npm run start`
 2. For a better dev experience, run your front repo too:
    - clone `https://github.com/NextUp-comedy/front-nc`
    - setup your `.env` file with:
      - <details>
         <summary>front env vars</summary>
       
         ```
            REACT_APP_CLIENT_ID = <XXXXXXXXXXX>
            REACT_APP_CLIENT_SECRET=XXXXXXXXXXXXXXXX
            SKIP_PREFLIGHT_CHECK=XXXX
         ```
        </details
    - `$ npm install` > `$ npm start`
----
#### Test this PR:
 1. Go to `api/v1/auth/signup/`:
      - Request "create user DTO"
       ```ts
       {
        "username": "your new name",
        "email": "valid_email@gmail.com",
        "password": "not_less_than_6_char"
       }
    ```
    - expected Response:
    ```ts
      {  "message": ""Check your email"   } // check your Email. " Spam / Primary / Promotions/ Updates / wait for half a min max if it wasn't there yet & make sure you've written your email correctly. 
     ```
2. See how the whole journey goes in this video 👇🏽 : 
   https://www.loom.com/share/bd8fbc4387b84a398cbf42d447d1b581

### CODE REVIEW

**Review by AymanSami98 — APPROVED** [review]

**Comment by Jood80** [comment]
An undesired behaviors have occurred here after the last merge.
Will need to tweak them then ask for your review again.

File: src/controllers/users/approveUser.ts [hunk 1]
```diff
@@ -8,6 +8,7 @@ import { checkExistence, constants } from '../../helpers';
 export default async (req: Request, res: Response, next: NextFunction)
 :Promise<void> => {
   const { userId } = req.params;
+  const redirectURL = `${config.server.clientURL}/`;
```
Comment (zaher-aa) [inline src/controllers/users/approveUser.ts:11]: The `/` after the clientURL is for what?
```suggestion
  const redirectURL = `${config.server.clientURL}`;
```

File: .eslintrc.json [hunk 2]
```diff
@@ -27,7 +27,21 @@
...
+      {
+        "types": {
+          "{}": false
+        },
+        "extendDefaults": true
+      }
+    ],
+    
+      "no-tabs": 0
+
```
Comment (zaher-aa) [inline .eslintrc.json:44]: Why did you add these rules?

**Review by zaher-aa — CHANGES_REQUESTED** [review]

File: src/config/email.ts [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+const envVarsSchema = Joi.object({
+  SENDGRID_ADMIN_EMAIL: Joi.string().required(),
+  SENDGRID_API_KEY: Joi.string().required(),
+  SENDGRID_VERIFICATION_TEMPLATE_ID: Joi.string().required(),
+  SENDGRID_RESET_PASSWORD_TEMPLATE_ID: Joi.string().required(),
+  SENDGRID_APPROVAL_TEMPLATE_ID: Joi.string().required(),
+  SENDGRID_REJECTION_TEMPLATE_ID: Joi.string().required(),
+  NEXTUP_COMEDY_SUPPORT_EMAIL: Joi.string().required(),
+}).unknown().required();
```
[the remark this answers is not in the document]
Comment (Jood80) [reply src/config/email.ts:12]: YUP, was just about to fix those in the next push 👍🏽

File: src/config/email.ts [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+  return {
+    SENDGRID_ADMIN_EMAIL: envVars.SENDGRID_ADMIN_EMAIL,
+    SENDGRID_API_KEY: envVars.SENDGRID_API_KEY,
+    SENDGRID_VERIFICATION_TEMPLATE_ID: envVars.SENDGRID_VERIFICATION_TEMPLATE_ID,
+    SENDGRID_RESET_PASSWORD_TEMPLATE_ID: envVars.SENDGRID_RESET_PASSWORD_TEMPLATE_ID,
+    SENDGRID_APPROVAL_TEMPLATE_ID: envVars.SENDGRID_APPROVAL_TEMPLATE_ID,
+    SENDGRID_REJECTION_TEMPLATE_ID: envVars.SENDGRID_REJECTION_TEMPLATE_ID,
+    NEXTUP_COMEDY_SUPPORT_EMAIL: envVars.NEXTUP_COMEDY_SUPPORT_EMAIL,
+  };
```
[the remark this answers is not in the document]
Comment (Jood80) [reply src/config/email.ts:28]: Absolutely right in case those are variables, but they ain't.

File: .eslintrc.json (hunk 2 above)
Comment (Jood80) [reply .eslintrc.json]: I'm keeping the `no-shadow` part only for now, as I ran into this problem in `src/interfaces/email.ts` file
image

The cute thing that I AM NOT declaring it somewhere else, which was confusing somehow.

**Review by RandSohail — APPROVED** [review]

**Review by zaher-aa — APPROVED** [review]

**Comment by zaher-aa** [comment]
👍

**Pull request merged**